sklearn/ensemble/_iforest.py

```diff
@@ -486,7 +486,18 @@ def _more_tags(self):
         }
 
 
-# Lookup table used below in _average_path_length() for small samples
+# Lookup table used below in _average_path_length() for small samples.
+# Since the average path length equals 2*(H(n)-1), with H(n) the nth
+# harmonic number, it can be calculated as Sum(2/k, k=2...n) for n>=2
+# and is zero for n<=1.
+# To achieve correct results to full rounding precision, the list below
+# can be generated using SymPy (https://www.sympy.org) as follows (r is
+# the result)
+#
+# from sympy import Sum, Symbol
+# k = Symbol('k')
+# r = [Sum(2/k, (k, 2, max(1, i))).evalf(22).round(19) for i in range(52)]
+#
 _average_path_length_small = np.array(
```

Comment:
Thanks for the snippet!

This seems to have a quadratic complexity, using a cumulative can make it linear. This can be done with numpy easily with a similar runtime.

```python
In [1]: import warnings, numpy as np
In [2]: def f(n):
   ...:     # Catch the RuntimeWarning due to the
   ...:     # division by 0 for the first term
   ...:     with warnings.catch_warnings():
   ...:         warnings.simplefilter("ignore")
   ...:         terms = 2 / np.arange(n)
   ...:     terms[0] = terms[1] = 0.
   ...:     # Compute the series of harmonic number
   ...:     _average_path_length_small = np.cumsum(terms)
In [3]: %timeit f(52)
8.57 µs ± 2.29 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)
In [4]: %timeit f(100000)
594 µs ± 2.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
In [5]: def a(): 
   ...:     # insert this inline definition of the array 
In [6]: %timeit a()
3.34 µs ± 7.51 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each
```

Importing `warnings` at the top of the file:

```suggestion
# Lookup table used below in _average_path_length() for small samples.
# Since the average path length equals 2*(H(n)-1), with H(n) the nth
# harmonic number, it can be calculated as Sum(2/k, k=2...n) for n>=2
# and is zero for n<=1.

# Catch the RuntimeWarning due to the division by 0 for the first term
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    terms = 2 / np.arange(n)
terms[0] = terms[1] = 0.
# Compute the series of harmonic number
_average_path_length_small = np.cumsum(terms)
```

What do you think?

Reply:
Hi @jjerphan,
thanks for the suggestion, I appreciate it. Sorry for the misunderstanding, of course I'm aware that the list can be generated in linear time, the code snippet in the comment was written to be as easily understandable as possible, not as an actual implementation. IMO we should keep the list, it's faster as you've shown (can't beat a lookup table on that ;-)) and I honestly don't think your code is more comprehensible to an uninitiated reader. Let me know what you think

Reply:
I think we need someone else to express their opinion on this point. :slightly_smiling_face: